### PR TITLE
Sort Actions workflow runs newest-first (Fixes #208)

### PR DESCRIPTION
**Pull request by Ayush7614**
## Summary
- Defensively sort Actions workflow runs by `created_at` descending (tie-break by `id`) on page-1 reload and after load-more appends so the list stays reverse-chronological regardless of API/page order
- Preserve selected run identity across post-append reorders
- Add unit and reducer regression tests for out-of-order loads, interleaved page appends, ties, and empty timestamps

Fixes #208

## Test plan
- [x] `cargo test --lib actions_tests`
- [x] `cargo test --lib sort_workflow_runs_newest_first`
- [x] `cargo clippy --lib -p jefe -- -D warnings`
- [ ] Manual: open Actions mode on a busy repo, confirm newest run is at top; scroll/load-more and confirm older runs stay below newer ones

## Summary by CodeRabbit

- **Improvements**
  - Actions workflow runs now appear in newest-first order, including after loading additional pages.
  - The selected run remains selected when newly loaded runs change the list order.
  - Run ordering is deterministic when timestamps are missing or identical.
  - Navigation and detail views now stay aligned with the displayed run order.

- **Tests**
  - Expanded coverage verifies sorting, pagination, selection stability, and Actions run detail navigation.

### CODE REVIEW

File: src/state/actions_load_ops.rs [hunk 1]
```diff
@@ -143,3 +150,19 @@ impl AppState {
...
+    let selected_id = list
+        .selected_index()
+        .and_then(|idx| list.items().get(idx).map(|run| run.id));
+    let mut runs = list.items().to_vec();
+    crate::actions_view::sort_workflow_runs_newest_first(&mut runs);
+    list.replace_items(runs);
+    if let Some(id) = selected_id {
+        list.set_selected_index(list.items().iter().position(|run| run.id == id));
+    }
+}
```
[the remark this answers is not in the document]
Comment (Ayush7614) [reply src/state/actions_load_ops.rs:173]: Addressed in the follow-up commit: added `PaginatedList::sort_by` for in-place reordering and switched `resort_actions_runs_preserving_selection` to use it, so load-more no longer clones the run list just to sort.

File: src/state/actions_tests.rs [hunk 2]
```diff
@@ -563,6 +567,87 @@ mod tests {
...
+        assert_eq!(
+            state.actions_state.selected_run().map(|r| r.id),
+            Some(2),
+            "visible reload selects the newest run at index 0"
+        );
+    }
+
+    /// Load-more appends re-sort so older pages slot below newer runs (issue #208).
+    #[test]
+    fn runs_page_loaded_resorts_so_older_appends_slot_below() {
```
[the remark this answers is not in the document]
Comment (Ayush7614) [reply src/state/actions_tests.rs]: Added `runs_load_paths_break_equal_created_at_ties_by_id_desc`: reload and page-append both assert equal `created_at` order as id-descending, and selection follows the same run id across the tied-timestamp resort.

File: src/actions_view.rs [hunk 3]
```diff
@@ -3,10 +3,35 @@
...
+pub fn cmp_workflow_runs_newest_first(a: &WorkflowRun, b: &WorkflowRun) -> Ordering {
+    b.created_at
+        .cmp(&a.created_at)
+        .then_with(|| b.id.cmp(&a.id))
+}
+
+/// Sort `runs` reverse-chronologically in place (most recent `created_at` first).
+pub fn sort_workflow_runs_newest_first(runs: &mut [WorkflowRun]) {
+    runs.sort_by(cmp_workflow_runs_newest_first);
+}
```
[the remark this answers is not in the document]
Comment (Ayush7614) [reply src/actions_view.rs]: Switched `cmp_workflow_runs_newest_first` and `sort_workflow_runs_newest_first` to `pub(crate)` since they are only consumed inside the crate.

**Review by acoliver — CHANGES_REQUESTED** [review]
## LLxprt Code review — request changes

> This review was performed and written by **LLxprt Code, an AI coding assistant**, at the repository owner's request. Although GitHub displays the repository owner's account as the author, the findings and wording below are the assistant's—not the account owner's personal review.

The core implementation appears behaviorally correct and appropriately scoped, but this PR is not merge-ready.

### Blockers

1. **Exact head fails formatting.** Running `cargo fmt --all --check` against commit `6ca9e231` exited 1 and produced changes in `src/actions_view.rs` around lines 190–238 and `src/state/actions_tests.rs` around lines 607–645.

2. **Required exact-head CI did not execute.** Actions run `29418432360` concluded `action_required` with zero jobs. Formatting, strict Clippy, policy checks, coverage, build, tests, and native Windows therefore remain unverified.

3. **Required TUI scenario coverage is absent.** This is UI-visible behavior, but neither `dev-docs/tmux-scenarios/actions-mode.json` nor `dev-docs/tmux-scenarios/actions-pr-filter.json` asserts newest-first initial ordering or ordering after load-more. The issue-delivery workflow requires scenario-first RED/GREEN evidence for UI-visible work.

4. **The branch is beyond the mainline-drift threshold.** PR head `6ca9e231` is 12 commits behind current remote main `32ffac73`. GitHub's mergeability result is based on a stale generated merge, so current-main compatibility must be established by integrating main and rerunning verification.

5. **Review state is incomplete.** The original CodeRabbit allocation concern was correctly fixed and resolved, but its attempted exact-head review was rate-limited and explicitly did not start. Two OCR threads also remain unresolved, although their requested changes appear substantively implemented.

6. **Auditable workflow/TDD evidence is incomplete.** There is no issue 208 acceptance/scope ledger, the first implementation commit introduced production code and tests together, and no TUI scenario-first RED evidence exists. Do not rewrite history to simulate TDD; add the missing ledger and execute a genuine RED/GREEN cycle for the TUI slice.

### Functional assessment

The implementation itself looks sound: it sorts initial and appended pages globally newest-first, is independent of API response order, preserves selection by workflow-run ID, defines deterministic equal-timestamp behavior, handles missing timestamps consistently, and uses an in-place sort rather than cloning the full list. Pure and reducer tests provide good lower-level coverage. No new panic, unsafe, suppression, dependency, or architectural-boundary concerns were found.

Scope is bounded at 5 files and +296/-3 lines, with no unrelated configuration or dependency changes.

### Independent verification

Against an isolated archive of exact head:

- `cargo fmt --all --check` — **failed**
- Source-size gate — passed
- Actions reducer tests — 31/31 passed
- Sorting-specific tests — 3/3 passed
- Complexity Clippy gate — passed
- Workspace build — passed
- Strict Clippy — failed on an unattributed configuration/baseline `large_stack_arrays` diagnostic
- Full tests reached 326 passing integration tests, then one repository-policy test failed because the archive lacked `.git`
- Coverage was incomplete for the same archive limitation

### Required path to approval

1. Add and execute the deterministic TUI ordering scenario with RED/GREEN evidence.
2. Apply rustfmt.
3. Integrate current main.
4. Run the complete exact-head local verification suite.
5. Get every required GitHub CI job to execute and pass, including Windows and coverage.
6. Complete an exact-head CodeRabbit review.
7. Triage and resolve the remaining OCR threads.

File: scripts/issue194-gh-shim.sh [hunk 4]
```diff
@@ -23,6 +30,26 @@ JSON
...
     cat <<'JSON'
 [{"databaseId":19411,"name":"build-linux-with-a-long-job-name","status":"completed","conclusion":"success","steps":[{"name":"Checkout fixture source","status":"completed","conclusion":"success","number":1},{"name":"Compile fixture application","status":"completed","conclusion":"success","number":2}]},{"databaseId":19412,"name":"test-suite-with-a-long-job-name","status":"completed","conclusion":"failure","steps":[{"name":"Run failing fixture tests","status":"completed","conclusion":"failure","number":1}]},{"databaseId":19413,"name":"publish-artifacts","status":"completed","conclusion":"skipped","steps":[]}]
+JSON
+    ;;
+  "run view --repo owner/actions-fixture 20801 --json attempt,conclusion,createdAt,databaseId,displayTitle,event,headBranch,headSha,name,number,startedAt,status,updatedAt,url,workflowDatabaseId,workflowName")
+    cat <<'JSON'
+{"databaseId":20801,"name":"Oldest Actions fixture run","headBranch":"issue208-old","headSha":"old208","number":101,"event":"push","status":"completed","conclusion":"success","workflowName":"CI","createdAt":"2026-07-01T00:00:00Z","updatedAt":"2026-07-01T00:01:00Z"}
```
[the remark this answers is not in the document]
Comment (acoliver) [reply scripts/issue194-gh-shim.sh:37]: LLxprt Code triage: Reject. The fixture includes every field consumed by GhWorkflowRunJson in src/github/actions.rs; serde ignores the additional fields requested from gh. The deterministic TUI scenario passes through this boundary. Expanding the fixture would not test application behavior.

File: scripts/issue194-gh-shim.sh [hunk 5]
```diff
@@ -23,6 +30,26 @@ JSON
...
+JSON
+    ;;
+  "run view --repo owner/actions-fixture 20801 --json jobs --jq .jobs")
+    cat <<'JSON'
+[]
+JSON
+    ;;
+  "run view --repo owner/actions-fixture 20800 --json attempt,conclusion,createdAt,databaseId,displayTitle,event,headBranch,headSha,name,number,startedAt,status,updatedAt,url,workflowDatabaseId,workflowName")
+    cat <<'JSON'
+{"databaseId":20800,"name":"Archived Actions fixture run","headBranch":"issue208-archive","headSha":"arc208","number":100,"event":"push","status":"completed","conclusion":"cancelled","workflowName":"CI","createdAt":"2026-06-01T00:00:00Z","updatedAt":"2026-06-01T00:01:00Z"}
```
[the remark this answers is not in the document]
Comment (acoliver) [reply scripts/issue194-gh-shim.sh]: LLxprt Code triage: Reject for the same parser-contract reason. The 20800 fixture supplies all fields deserialized by GhWorkflowRunJson, and omitted requested fields are ignored. The interleaved pagination scenario passes and proves the intended ordering path.

File: src/state/actions_load_ops.rs [hunk 6]
```diff
@@ -143,3 +151,23 @@ impl AppState {
         self.actions_state.list.begin_reload(identity, request_id);
     }
 }
+
+fn cmp_workflow_runs_newest_first(a: &WorkflowRun, b: &WorkflowRun) -> Ordering {
+    b.created_at
+        .cmp(&a.created_at)
+        .then_with(|| b.id.cmp(&a.id))
+}
```
[the remark this answers is not in the document]
Comment (acoliver) [reply src/state/actions_load_ops.rs:159]: LLxprt Code triage: Reject. GitHub Actions workflow-run created_at is an RFC 3339 UTC timestamp in this boundary contract, and the repository already represents it as a string. Supporting hypothetical offset or precision variants would expand issue 208 beyond the API contract and add a dependency/abstraction without evidence.

File: src/state/actions_load_ops.rs [hunk 7]
```diff
@@ -63,6 +67,10 @@ impl AppState {
         };
         let outcome = self.actions_state.list.accept_page(result);
         if matches!(outcome, AcceptOutcome::Applied | AcceptOutcome::Empty) {
+            // Re-sort the full list after append so older pages slot below newer
+            // runs even when API/page order is wrong (issue #208). Preserve the
+            // selected run by id because indices may shift.
+            resort_actions_runs_preserving_selection(&mut self.actions_state.list);
             self.actions_state.error = None;
         }
```
[the remark this answers is not in the document]
Comment (acoliver) [reply src/state/actions_load_ops.rs:75]: LLxprt Code triage: Reject as non-actionable. Sorting an unchanged list on the rare empty-page completion is bounded and keeps the accepted-result invariant simple; changing this branch would be a micro-optimization without measurable benefit or correctness impact.

**Review by acoliver — APPROVED** [review]
## LLxprt Code review — approved

> This approval was performed and written by **LLxprt Code, an AI coding assistant**, at the repository owner's request. Although GitHub displays the repository owner's account as the author, this is the assistant's review.

The exact-head candidate at `29c5307` is ready to merge.

- Issue 208 behavior is covered at reducer and TUI levels, including discriminating interleaved pagination and selection preservation.
- Ordering ownership respects the architecture boundary.
- The API surface is appropriately crate-private.
- Full local `make ci-check` passed.
- All required exact-head GitHub checks passed, including coverage and native Windows/psmux.
- CodeRabbit completed with no actionable findings.
- OCR findings were evaluated, answered, and resolved.
- All review threads are resolved.
- Current main is integrated with correct ancestry.
- GitHub reports the PR mergeable and conflict-free.

The maintainer clarified that scenario-first TDD is a preferred contributor methodology rather than a merge gate for external contributions. The documented sequencing deviation is therefore accepted and does not block this PR.
